### ISSUE TITLE
Incorrect constructor creates a sink case

### ISSUE THREAD

**Issue by yacinehmito** [idris-lang/Idris-dev#3153]
Consider this snippet.

``` idris
data Vehicule = Bus | Car

wheels : Vehicule -> Nat
wheels ObviousTypo = 72
```

When loaded no error is thrown. It parses and typechecks fine.
In the REPL, doing `wheels Bus` gives 72. I think that the unknown constructor acts as a sink case.
This also happens if you declare the function total.

The throwing of a "No such variable" error (or similar) is the expected behavior.

My version is 0.11.1-2

**Comment by ahmadsalim** [comment]
This is because `ObviousTypo` is treated as a pattern variable which is perfectly fine.
I believe there has been discussions on whether such code as you presented should be an error or give a warning.

**Comment by gpyh** [comment]
Oh I see. It's because `wheels x` would have been entirely valid and Idris allows arguments starting by a capital letter.

After some thoughts I am not so fond of discriminating between capitalized and not capitalized identifier. I could see myself using capital letters as argument of some function.
Say I am writing a function to multiply a matrix and a vector, I would like to write `mult M u`.

However, Idris could warn/raise an error if an argument is not used. If a constructor was intended to be used, the user can fix the type. If it was just a useless binding, the user can replace it with an underscore.

**Comment by jfdm** [comment]
@gpyh IIRC, the need for discriminating between an identifier's case is not when calling a function but during pattern matching and in type signatures. You example of `mult M u` is fine, and perfectly valid. It is only when you are defining the type signature and are providing the function's pattern variables that you need to be more careful with case.

**Comment by gpyh** [comment]
I made a typo on the word typo. I wrote type instead. The irony...

Now, say that I have a type `data Symbol = A | B`. I want to write a function as such

```idris
next : Symbol -> Symbol
next A = B
next B = A
```

But I made a typo! I actually wrote:

```idris
next : Symbol -> Symbol
next A = B
next C = A
```

I would very much like Idris to tell me that I made a type and that `C` is not a valid constructor.
If I understand well, the declaration is fine because `C` is a valid argument name. As such, passing to `next` any `Symbol` that is not of `A` constructor will return `A` (which is fine given my example but you see the point).

However, Idris might know that my `C` binding is actually never used in the RHS. If the compiler warns me, I can see that I made a typo and change `C` to `B`. If it was my indeed intent to type `C`, I can replace it by an underscore. This will preserve the behavior I expected and suppress the warning altogether.

One big hurdle of this is that it doesn't play very well with big constructors (mostly records). The interactive editing takes care of writing the declaration for you. It can be annoying to be compelled to replace the unused arguments with sinks, especially if you're in the process of figuring out which arguments you need.

**Comment by edwinb** [comment]
I have wondered for some time if the right thing to do here is to have the same binding rules as in type declarations - namely, it's only a pattern variable if it begins with a lower case letter. This has been something I've tried to avoid, but for the sake of useful error messages and unsurprising behaviour, it might be the right thing to do.

I know that people would prefer not to have such a rule (I have tried to avoid it for years too after all), but any kind of semantic rule we've had to try to distinguish always turns out to be complicated and have all kinds of weird edge cases. Keeping things simple might be best.
